File: app/src/main/java/de/sysitprep/app/ui/quiz/QuizViewModel.kt
package de.sysitprep.app.ui.quiz

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import de.sysitprep.app.ExamForgeApp
import de.sysitprep.app.data.model.*
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

class QuizViewModel(application: Application) : AndroidViewModel(application) {

    private val repo = (application as ExamForgeApp).repository

    // ── State ─────────────────────────────────────────────────────────────────

    private val _questions = MutableLiveData<List<Question>>()
    val questions: LiveData<List<Question>> = _questions

    private val _currentIndex = MutableLiveData(0)
    val currentIndex: LiveData<Int> = _currentIndex

    private val _currentQuestion = MutableLiveData<Question?>()
    val currentQuestion: LiveData<Question?> = _currentQuestion

    private val _selectedIndices = MutableLiveData<Set<Int>>(emptySet())
    val selectedIndices: LiveData<Set<Int>> = _selectedIndices

    private val _isAnswerChecked = MutableLiveData(false)
    val isAnswerChecked: LiveData<Boolean> = _isAnswerChecked

    private val _isCorrect = MutableLiveData(false)
    val isCorrect: LiveData<Boolean> = _isCorrect

    private val _quizFinished = MutableLiveData(false)
    val quizFinished: LiveData<Boolean> = _quizFinished

    private val _sessionId = MutableLiveData<Long>(-1)
    val sessionId: LiveData<Long> = _sessionId

    // For exam simulation: countdown timer
    private val _timerSeconds = MutableLiveData<Int>()
    val timerSeconds: LiveData<Int> = _timerSeconds

    private val _timeExpired = MutableLiveData(false)
    val timeExpired: LiveData<Boolean> = _timeExpired

    private var timerJob: Job? = null
    private var questionStartTime = System.currentTimeMillis()
    private val collectedAnswers = mutableListOf<UserAnswer>()
    private var sessionType = SessionType.QUICK_QUIZ
    private var examTimeLimitSeconds = 0
    private var activeSessionId = -1L

    // ── Setup ─────────────────────────────────────────────────────────────────

    fun init(
        type: SessionType,
        lernfeld: Int,
        fachrichtung: Fachrichtung,
        examType: ExamType,
        isExamMode: Boolean = false
    ) {
        sessionType = type
        viewModelScope.launch {
            // Wait for seeding to complete before querying
            (getApplication<ExamForgeApp>()).seedingComplete.first { it }
            val questionList = when (type) {
                SessionType.LERNFELD ->
                    repo.getQuestionsForLernfeld(lernfeld, fachrichtung, 20).shuffled()
                SessionType.EXAM_SIMULATION ->
                    repo.getQuestionsForExam(examType, fachrichtung, 40).shuffled()
                SessionType.QUICK_QUIZ ->
                    repo.getRandomQuestions(fachrichtung, 10).shuffled()
                SessionType.BOOKMARK_REVIEW ->
                    emptyList() // populated via bookmark flow
            }

            if (questionList.isEmpty()) return@launch

            _questions.postValue(questionList)
            _currentIndex.postValue(0)
            _currentQuestion.postValue(questionList.firstOrNull())
            questionStartTime = System.currentTimeMillis()

            // Create session
            val session = Session(
                sessionType = type,
                fachrichtung = fachrichtung,
                examType = if (type == SessionType.EXAM_SIMULATION) examType else null,
                lernfeld = if (type == SessionType.LERNFELD) lernfeld else null,
                totalQuestions = questionList.size
            )
            activeSessionId = repo.startSession(session)
            _sessionId.postValue(activeSessionId)

            // Start exam timer if needed
            if (type == SessionType.EXAM_SIMULATION) {
                examTimeLimitSeconds = 90 * 60 // 90 minutes
                startTimer()
            }
        }
    }

    // ── Answer Actions ────────────────────────────────────────────────────────

    fun toggleOption(index: Int) {
        if (_isAnswerChecked.value == true) return
        val current = _selectedIndices.value?.toMutableSet() ?: mutableSetOf()
        val question = _currentQuestion.value ?: return
        if (question.correctAnswerIndices.size == 1) {
            // Single choice: only one can be selected
            current.clear()
            current.add(index)
        } else {
            // Multiple choice: toggle
            if (current.contains(index)) current.remove(index) else current.add(index)
        }
        _selectedIndices.value = current
    }

    fun checkAnswer() {
        val question = _currentQuestion.value ?: return
        val selected = _selectedIndices.value ?: return
        val correctSet = question.correctAnswerIndices.toSet()
        val isCorrect = selected == correctSet
        _isCorrect.value = isCorrect
        _isAnswerChecked.value = true

        val timeSpent = ((System.currentTimeMillis() - questionStartTime) / 1000).toInt()
        val answer = UserAnswer(
            sessionId = activeSessionId,
            questionId = question.id,
            selectedIndices = selected.toList(),
            isCorrect = isCorrect,
            timeSpentSeconds = timeSpent
        )
        collectedAnswers.add(answer)
    }

    fun nextQuestion() {
        val list = _questions.value ?: return
        val nextIndex = (_currentIndex.value ?: 0) + 1
        if (nextIndex >= list.size) {
            finishQuiz()
        } else {
            _currentIndex.value = nextIndex
            _currentQuestion.value = list[nextIndex]
            _selectedIndices.value = emptySet()
            _isAnswerChecked.value = false
            questionStartTime = System.currentTimeMillis()
        }
    }

    fun skipQuestion() {
        val question = _currentQuestion.value ?: return
        val timeSpent = ((System.currentTimeMillis() - questionStartTime) / 1000).toInt()
        collectedAnswers.add(
            UserAnswer(
                sessionId = activeSessionId,
                questionId = question.id,
                selectedIndices = emptyList(),
                isCorrect = false,
                timeSpentSeconds = timeSpent
            )
        )
        nextQuestion()
    }

    fun toggleBookmark() {
        val question = _currentQuestion.value ?: return
        viewModelScope.launch {
            repo.setBookmark(question.id, !question.isBookmarked)
        }
    }

    // ── Finish ────────────────────────────────────────────────────────────────

    private fun finishQuiz() {
        timerJob?.cancel()
        viewModelScope.launch {
            repo.saveAnswers(collectedAnswers)
            val correctCount = collectedAnswers.count { it.isCorrect }
            val totalDuration = collectedAnswers.sumOf { it.timeSpentSeconds }
            repo.finishSession(
                Session(
                    id = activeSessionId,
                    sessionType = sessionType,
                    fachrichtung = Fachrichtung.SI, // replace with actual
                    totalQuestions = _questions.value?.size ?: 0,
                    correctAnswers = correctCount,
                    durationSeconds = totalDuration,
                    completedAt = System.currentTimeMillis()
                )
            )
            _quizFinished.postValue(true)
        }
    }

    fun getResult(): QuizResult {
        val total = _questions.value?.size ?: 0
        val correct = collectedAnswers.count { it.isCorrect }
        val wrong = collectedAnswers.count { !it.isCorrect && it.selectedIndices.isNotEmpty() }
        val skipped = collectedAnswers.count { it.selectedIndices.isEmpty() }
        val duration = collectedAnswers.sumOf { it.timeSpentSeconds }
        val score = if (total == 0) 0 else (correct * 100 / total)
        val questionsList = _questions.value ?: emptyList()
        return QuizResult(
            sessionId = activeSessionId,
            totalQuestions = total,
            correctAnswers = correct,
            wrongAnswers = wrong,
            skippedAnswers = skipped,
            timeSeconds = duration,
            scorePercent = score,
            passed = score >= 50,
            questionResults = collectedAnswers.mapIndexed { i, a ->
                val q = questionsList.getOrNull(i) ?: return@mapIndexed null
                QuestionResult(
                    questionId = q.id,
                    questionText = q.questionText,
                    options = q.options,
                    correctIndices = q.correctAnswerIndices,
                    selectedIndices = a.selectedIndices,
                    isCorrect = a.isCorrect,
                    explanation = q.explanation
                )
            }.filterNotNull()
        )
    }

    // ── Timer (Exam Mode) ─────────────────────────────────────────────────────

    private fun startTimer() {
        _timerSeconds.value = examTimeLimitSeconds
        timerJob = viewModelScope.launch {
            var remaining = examTimeLimitSeconds
            while (remaining > 0) {
                delay(1000)
                remaining--
                _timerSeconds.postValue(remaining)
            }
            _timeExpired.postValue(true)
            finishQuiz()
        }
    }

    override fun onCleared() {
        super.onCleared()
        timerJob?.cancel()
    }
}

File: ExamForge-iOS/ExamForge/ViewModels/QuizViewModel.swift
import Foundation
import Combine

@MainActor
final class QuizViewModel: ObservableObject {

    // MARK: - Published State

    @Published var questions: [Question] = []
    @Published var currentIndex: Int = 0
    @Published var currentQuestion: Question?
    @Published var selectedIndices: Set<Int> = []
    @Published var isAnswerChecked: Bool = false
    @Published var isCorrect: Bool = false
    @Published var quizFinished: Bool = false
    @Published var sessionId: Int64 = -1
    @Published var timerSeconds: Int = 0
    @Published var isLoading: Bool = true

    private let repository: QuestionRepository
    private var params: QuizParams
    private var timerTask: Task<Void, Never>?
    private var questionStartTime: Date = Date()
    private var collectedAnswers: [UserAnswer] = []
    private var activeSessionId: Int64 = -1
    private let examTimeLimitSeconds = 90 * 60     // 5400 s

    var isExamMode: Bool { params.sessionType == .EXAM_SIMULATION }

    init(params: QuizParams, repository: QuestionRepository) {
        self.params = params
        self.repository = repository
    }

    // MARK: - Initialisation

    func start() {
        Task {
            isLoading = true
            let questionList: [Question]
            switch params.sessionType {
            case .LERNFELD:
                questionList = await repository.getQuestionsForLernfeld(lernfeld: params.lernfeld, fachrichtung: params.fachrichtung, limit: 20).shuffled()
            case .EXAM_SIMULATION:
                questionList = await repository.getQuestionsForExam(examType: params.examType, fachrichtung: params.fachrichtung, limit: 40).shuffled()
            case .QUICK_QUIZ:
                questionList = await repository.getRandomQuestions(fachrichtung: params.fachrichtung, limit: 10).shuffled()
            case .BOOKMARK_REVIEW:
                questionList = (await repository.getBookmarkedQuestions()).shuffled()
            }

            guard !questionList.isEmpty else {
                isLoading = false
                return
            }

            questions = questionList
            currentIndex = 0
            currentQuestion = questionList.first
            questionStartTime = Date()

            let session = Session(
                id: PersistenceController.shared.generateId(),
                sessionType: params.sessionType,
                fachrichtung: params.fachrichtung,
                examType: params.sessionType == .EXAM_SIMULATION ? params.examType : nil,
                lernfeld: params.sessionType == .LERNFELD ? params.lernfeld : nil,
                totalQuestions: questionList.count,
                correctAnswers: 0, durationSeconds: 0, completedAt: nil,
                startedAt: Date(), questionIds: []
            )
            activeSessionId = await repository.startSession(session)
            sessionId = activeSessionId

            if params.sessionType == .EXAM_SIMULATION {
                timerSeconds = examTimeLimitSeconds
                startTimer()
            }
            isLoading = false
        }
    }

    // MARK: - User Actions

    func toggleOption(_ index: Int) {
        guard !isAnswerChecked else { return }
        if currentQuestion?.correctAnswerIndices.count == 1 {
            selectedIndices = [index]
        } else {
            var updated = selectedIndices
            if updated.contains(index) { updated.remove(index) } else { updated.insert(index) }
            selectedIndices = updated
        }
    }

    func checkAnswer() {
        guard let question = currentQuestion else { return }
        let correct = question.correctAnswerIndices.count == 1
            ? selectedIndices == Set(question.correctAnswerIndices)
            : selectedIndices == Set(question.correctAnswerIndices)

        isCorrect = correct
        isAnswerChecked = true

        let timeSpent = Int(Date().timeIntervalSince(questionStartTime))
        collectedAnswers.append(UserAnswer(
            id: PersistenceController.shared.generateId(),
            sessionId: activeSessionId,
            questionId: question.dbId,
            selectedIndices: Array(selectedIndices),
            isCorrect: correct,
            timeSpentSeconds: timeSpent,
            answeredAt: Date()
        ))
    }

    func nextQuestion() {
        let nextIdx = currentIndex + 1
        if nextIdx >= questions.count {
            finishQuiz()
        } else {
            currentIndex = nextIdx
            currentQuestion = questions[nextIdx]
            selectedIndices = []
            isAnswerChecked = false
            questionStartTime = Date()
        }
    }

    func skipQuestion() {
        guard let question = currentQuestion else { return }
        let timeSpent = Int(Date().timeIntervalSince(questionStartTime))
        collectedAnswers.append(UserAnswer(
            id: PersistenceController.shared.generateId(),
            sessionId: activeSessionId,
            questionId: question.dbId,
            selectedIndices: [],
            isCorrect: false,
            timeSpentSeconds: timeSpent,
            answeredAt: Date()
        ))
        nextQuestion()
    }

    func toggleBookmark() {
        guard let question = currentQuestion else { return }
        Task {
            await repository.setBookmark(id: question.dbId, bookmarked: !question.isBookmarked)
            if let idx = questions.firstIndex(where: { $0.dbId == question.dbId }) {
                var updated = questions[idx]
                updated.isBookmarked.toggle()
                questions[idx] = updated
                currentQuestion = questions[currentIndex]
            }
        }
    }

    // MARK: - Private

    private func finishQuiz() {
        timerTask?.cancel()
        Task {
            await repository.saveAnswers(collectedAnswers)
            let correctCount = collectedAnswers.filter { $0.isCorrect }.count
            let totalDuration = collectedAnswers.reduce(0) { $0 + $1.timeSpentSeconds }
            let session = Session(
                id: activeSessionId,
                sessionType: params.sessionType,
                fachrichtung: params.fachrichtung,
                examType: params.examType,
                lernfeld: params.lernfeld == -1 ? nil : params.lernfeld,
                totalQuestions: questions.count,
                correctAnswers: correctCount,
                durationSeconds: totalDuration,
                completedAt: Date(),
                startedAt: Date(),
                questionIds: []
            )
            await repository.finishSession(session)
            quizFinished = true
        }
    }

    private func startTimer() {
        timerTask = Task {
            while timerSeconds > 0 {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                if Task.isCancelled { return }
                timerSeconds -= 1
            }
            finishQuiz()
        }
    }

    func cancelTimer() {
        timerTask?.cancel()
    }
}
